src/wp-includes/blocks.php

```diff
@@ -2083,3 +2083,98 @@ function _wp_footnotes_force_filtered_html_on_import_filter( $arg ) {
+ *
+ * ### Example
+ *
+ * The "bindings" property for an Image block might look like this:
+ *
+ * ```json
+ * {
+ *   "metadata": {
+ *     "bindings": {
+ *       "title": {
+ *         "source": {
+ *           "name": "post_meta",
+ *           "attributes": { "value": "text_custom_field" }
+ *         }
+ *       },
+ *       "url": {
+ *         "source": {
+ *           "name": "post_meta",
+ *           "attributes": { "value": "url_custom_field" }
+ *         }
+ *       }
+ *     }
+ *   }
+ * }
+ * ```
+ *
+ * The above example will replace the `title` and `url` attributes of the Image
+ * block with the values of the `text_custom_field` and `url_custom_field` post meta.
+ *
+ * @access private
+ * @since 6.5.0
+ *
+ * @param string   $block_content Block content.
+ * @param array    $block The full block, including name and attributes.
+ * @param WP_Block $block_instance The block instance.
+ */
+function _process_block_bindings( $block_content, $block, $block_instance ) {
+
+	// Allowed blocks that support block bindings.
+	// TODO: Look for a mechanism to opt-in for this. Maybe adding a property to block attributes?
+	$allowed_blocks = array(
+		'core/paragraph' => array( 'content' ),
+		'core/heading'   => array( 'content' ),
+		'core/image'     => array( 'url', 'title', 'alt' ),
+		'core/button'    => array( 'url', 'text' ),
+	);
+
+	// If the block doesn't have the bindings property or isn't one of the allowed block types, return.
+	if ( ! isset( $block['attrs']['metadata']['bindings'] ) || ! isset( $allowed_blocks[ $block_instance->name ] ) ) {
+		return $block_content;
+	}
+
+	$block_bindings_sources = wp_block_bindings_get_sources();
+	$modified_block_content = $block_content;
+	foreach ( $block['attrs']['metadata']['bindings'] as $binding_attribute => $binding_source ) {
+
+		// If the attribute is not in the list, process next attribute.
+		if ( ! in_array( $binding_attribute, $allowed_blocks[ $block_instance->name ], true ) ) {
+			continue;
+		}
+		// If no source is provided, or that source is not registered, process next attribute.
+		if ( ! isset( $binding_source['source'] ) || ! isset( $binding_source['source']['name'] ) || ! isset( $block_bindings_sources[ $binding_source['source']['name'] ] ) ) {
+			continue;
+		}
+
+		$source_callback = $block_bindings_sources[ $binding_source['source']['name'] ]['apply'];
+		// Get the value based on the source.
+		if ( ! isset( $binding_source['source']['attributes'] ) ) {
+			$source_args = array();
+		} else {
+			$source_args = $binding_source['source']['attributes'];
+		}
+		$source_value = $source_callback( $source_args, $block_instance, $binding_attribute );
+		// If the value is null, process next attribute.
+		if ( is_null( $source_value ) ) {
+			continue;
+		}
+
+		// Process the HTML based on the block and the attribute.
+		$modified_block_content = wp_block_bindings_replace_html( $modified_block_content, $block_instance->name, $binding_attribute, $source_value );
```

Comment:
When `_process_block_bindings` becomes a private method of the `WP_Block` class, then `wp_block_bindings_replace_html` could probably also be moved there as another private method. Is there anything specific there that makes it better suited to be located in the `WP_Block_Bindings` class?

Reply:
You're right. I don't see a reason to expose it now. If it's needed for something later we can expose it then.